### PR TITLE
Implement RFC 57

### PR DESCRIPTION
**Pull request by cdisselkoen**
## Issue #, if available

[RFC 57](cedar-policy/rfcs#57)

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A bug fix or other functionality change requiring a patch to `cedar-policy`.

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by john-h-kastner-aws — APPROVED** [review]
Less code. Love it

File: cedar-policy-core/src/est/err.rs [hunk 1]
```diff
@@ -80,6 +80,10 @@ pub enum FromJsonError {
     #[error("invalid entity type: {0}")]
     #[diagnostic(transparent)]
     InvalidEntityType(ParseErrors),
+    /// Errors constructing AST expressions
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    ExprConstructionError(#[from] ast::ExprConstructionError),
```
Comment (john-h-kastner-aws) [inline cedar-policy-core/src/est/err.rs:86]: What's this doing? Feels orthogonal to the rest of the PR.

**Comment by john-h-kastner-aws** [comment]
Also, this should fix #619, right?

**Comment by cdisselkoen** [comment]
Yes, this fixes #619.  I should add test cases for the tests in @andrewmwells-amazon 's comment there

File: cedar-policy-core/src/est/err.rs (hunk 1 above)
Comment (cdisselkoen) [reply cedar-policy-core/src/est/err.rs]: Removed.

File: cedar-policy-core/src/parser/cst_to_ast.rs [hunk 2]
```diff
@@ -1473,53 +1473,13 @@ impl Node<Option<cst::Mult>> {
...
-                    .next()
-                    .expect("already checked that it's not empty");
-                Some(ExprOrSpecial::Expr {
-                    expr: construct_expr_mul(nonconstantint, constantints, &self.loc),
-                    loc: self.loc.clone(),
-                })
-            }
+            // remove the opcodes -- from here on we assume they're all `Times`,
+            // having checked above that this is the case
+            let more = more.into_iter().map(|(_, e)| e);
```
Comment (aaronjeline) [inline cedar-policy-core/src/parser/cst_to_ast.rs:1478]: We could do this via a `filter_map` instead of having the invariant

**Review by aaronjeline — APPROVED** [review]
Less code good code

**Review by khieta — APPROVED** [review]
Lgtm. But this seems more appropriate for a minor than a patch, since it's adding support for a new class of expressions.

File: cedar-policy-core/src/parser/cst_to_ast.rs (hunk 2 above)
Comment (cdisselkoen) [reply cedar-policy-core/src/parser/cst_to_ast.rs]: Refactored.  Possible small behavior change in the case where `first.into_expr` throws an error and there was a `/` or `%` -- we used to prefer the `first.into_expr` error and now we'll prefer the `/` or `%` error.  But I think that's acceptable.

**Review by aaronjeline — APPROVED** [review]

**Review by shaobo-he-aws — APPROVED** [review]
LGTM.

**Review by aaronjeline — APPROVED** [review]

**Review by shaobo-he-aws — APPROVED** [review]

**Review by aaronjeline — APPROVED** [review]

**Review by aaronjeline — APPROVED** [review]

**Review by aaronjeline — APPROVED** [review]
